bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -893,20 +896,28 @@ private static boolean checkIsMapType(MapValue sourceVal, BRecordType sourceType
     }
 
     private static boolean checkIsXMLType(Type sourceType, Type targetType, List<TypePair> unresolvedTypes) {
-        if (sourceType.getTag() == TypeTags.FINITE_TYPE_TAG) {
+        int sourceTag = sourceType.getTag();
+        if (sourceTag == TypeTags.FINITE_TYPE_TAG) {
             return isFiniteTypeMatch((BFiniteType) sourceType, targetType);
         }
+
         BXmlType target = ((BXmlType) targetType);
-        if (sourceType.getTag() == TypeTags.XML_TAG) {
+        if (sourceTag == TypeTags.XML_TAG) {
             Type targetConstraint = target.constraint;
             // TODO: Revisit and check why xml<xml<constraint>>> on chained iteration
             while (target.constraint.getTag() == TypeTags.XML_TAG) {
                 target = (BXmlType) target.constraint;
                 targetConstraint = target.constraint;
             }
-            return checkIsType(((BXmlType) sourceType).constraint, targetConstraint,
-                               unresolvedTypes);
-        } else if (TypeTags.isXMLTypeTag(sourceType.getTag())) {
+            BXmlType source = (BXmlType) sourceType;
+            if (source.constraint.getTag() == TypeTags.NEVER_TAG) {
+                return targetConstraint.getTag() == TypeTags.UNION_TAG ||
```

Comment:
We need to check the members of the union right? Just being a union is not enough?

Reply:
Can there be types like xml<'xml:Text | `xml: Element> ? 
if not then this is correct right. constraint is a union only when it is of type xml